// vue-router-dom/dist-deno/index.ts
export * from './components/index'

export { useBlocker } from './hooks/useBlocker'
export { useHref } from './hooks/useHref'
// export { useInRouterContext } from './hooks/useInRouterContext'
export { useLocation } from './hooks/useLocation'
export { useMatch } from './hooks/useMatch'
export { useNavigate } from './hooks/useNavigate'
export { useOutlet } from './hooks/useOutlet'
export { useParams } from './hooks/useParams'
export { usePrompt } from './hooks/usePrompt'
export { useResolvedPath } from './hooks/useResolvedPath'
export { useRoutes } from './hooks/useRoutes'
export { useSearchParams } from './hooks/useSearchParams'

export { createRoutesFromArray } from './api/createRoutesFromArray'
export { createRoutesFromChildren } from './api/createRoutesFromChildren'
export { generatePath } from './api/generatePath'
export { matchRoutes } from './api/matchRoutes'
export { matchPath } from './api/matchPath'
export { resolvePath } from './api/resolvePath'

export * from './api/install'
export * from './api/keys'
export * from './api/types'

export * from 'history'
